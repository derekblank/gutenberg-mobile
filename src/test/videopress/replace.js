/**
 * WordPress dependencies
 */
import { dispatch } from '@wordpress/data';
import { store as coreStore } from '@wordpress/core-data';
import { requestImageUploadCancelDialog } from '@wordpress/react-native-bridge';

/**
 * External dependencies
 */
import {
	dismissModal,
	fireEvent,
	getBlock,
	getEditorHtml,
	initializeEditor,
	setupCoreBlocks,
	setupMediaUpload,
	setupMediaPicker,
	setupPicker,
	setupApiFetch,
} from 'test/helpers';

/**
 * Internal dependencies
 */
import {
	registerJetpackBlocks,
	setupJetpackEditor,
} from '../../jetpack-editor-setup';
import {
	expectShowMediaOptions,
	generateFetchMocks,
	generateBlockHTML,
	generateUploadMetadata,
	generateRemoteMedia,
	generateLocalMedia,
	generateLibraryMedia,
	sendWebViewMessage,
} from './local-helpers/utils';
import { MEDIA_OPTIONS } from './local-helpers/constants';

jest.mock( '@wordpress/api-fetch' );

const GUID_INITIAL_VIDEO = 'OLDxGUID';
const GUID_REPLACE_VIDEO = 'NEWxGUID';

setupCoreBlocks();

beforeAll( () => {
	// Register VideoPress block
	setupJetpackEditor( {
		blogId: 1,
		isJetpackActive: true,
	} );
	registerJetpackBlocks( {
		capabilities: { videoPressBlock: true },
	} );
} );

beforeEach( () => {
	// Invalidate `getEmbedPreview` resolutions to avoid
	// caching the preview for the same VideoPress GUID.
	dispatch( coreStore ).invalidateResolutionForStoreSelector(
		'getEmbedPreview'
	);

	// Mock API responses for initial block.
	setupApiFetch(
		generateFetchMocks( {
			guid: GUID_INITIAL_VIDEO,
			metadata: {
				title: 'Video to be replaced',
			},
		} )
	);
} );

describe( 'VideoPress block - Replace', () => {
	it( 'displays media options picker when replacing the video', async () => {
		// Mock API responses for default VideoPress GUID
		setupApiFetch( generateFetchMocks() );

		const screen = await initializeEditor( {
			initialHtml: generateBlockHTML(),
		} );
		const { getByLabelText } = screen;

		// Block is visible
		const block = await getBlock( screen, 'VideoPress' );
		expect( block ).toBeVisible();
		fireEvent.press( block );

		// Replace video
		fireEvent.press( getByLabelText( 'Edit video' ) );

		// Observe that media options picker is displayed
		expectShowMediaOptions( screen, {
			title: 'Replace video',
			options: MEDIA_OPTIONS,
		} );
	} );

	it( 'replaces with a local video', async () => {
		const screen = await initializeEditor( {
			initialHtml: generateBlockHTML( { guid: GUID_INITIAL_VIDEO } ),
		} );
		const { getByLabelText, getByTestId } = screen;

		// Block is visible
		const block = await getBlock( screen, 'VideoPress' );
		expect( block ).toBeVisible();
		fireEvent.press( block );

		expect( getEditorHtml() ).toMatchSnapshot( 'before replacing video' );

		// Setup pickers and media upload
		const { notifySucceedState } = setupMediaUpload();
		const { expectMediaPickerCall, mediaPickerCallback } =
			setupMediaPicker();
		const { selectOption } = setupPicker( screen, MEDIA_OPTIONS );

		// Mock API responses for new block
		setupApiFetch(
			generateFetchMocks( {
				guid: GUID_REPLACE_VIDEO,
				metadata: {
					title: 'Local video from device',
				},
			} )
		);

		// Replace video
		fireEvent.press( getByLabelText( 'Edit video' ) );

		// Upload video from device
		selectOption( 'Choose from device' );
		expectMediaPickerCall( 'DEVICE_MEDIA_LIBRARY', [ 'video' ], false );
		await mediaPickerCallback( generateLocalMedia() );

		// Upload succeed
		await notifySucceedState( {
			...generateRemoteMedia(),
			...generateUploadMetadata( GUID_REPLACE_VIDEO ),
		} );

		expect( getByTestId( 'videopress-player' ) ).toBeVisible();
		expect( getEditorHtml() ).toMatchSnapshot( 'after replacing video' );
	} );

	it( 'takes a new video', async () => {
		const screen = await initializeEditor( {
			initialHtml: generateBlockHTML( { guid: GUID_INITIAL_VIDEO } ),
		} );
		const { getByLabelText, getByTestId } = screen;

		// Block is visible
		const block = await getBlock( screen, 'VideoPress' );
		expect( block ).toBeVisible();
		fireEvent.press( block );

		expect( getEditorHtml() ).toMatchSnapshot( 'before replacing video' );

		// Setup pickers and media upload
		const { notifySucceedState } = setupMediaUpload();
		const { expectMediaPickerCall, mediaPickerCallback } =
			setupMediaPicker();
		const { selectOption } = setupPicker( screen, MEDIA_OPTIONS );

		// Mock API responses for new block
		setupApiFetch(
			generateFetchMocks( {
				guid: GUID_REPLACE_VIDEO,
				metadata: {
					title: 'Recorded within the editor',
				},
			} )
		);

		// Replace video
		fireEvent.press( getByLabelText( 'Edit video' ) );

		// Take a video and upload it
		selectOption( 'Take a Video' );
		expectMediaPickerCall( 'DEVICE_CAMERA', [ 'video' ], false );
		await mediaPickerCallback( generateLocalMedia() );

		// Upload succeed
		await notifySucceedState( {
			...generateRemoteMedia(),
			...generateUploadMetadata( GUID_REPLACE_VIDEO ),
		} );

		expect( getByTestId( 'videopress-player' ) ).toBeVisible();
		expect( getEditorHtml() ).toMatchSnapshot( 'after replacing video' );
	} );

	it( 'replaces with video from media library', async () => {
		const screen = await initializeEditor( {
			initialHtml: generateBlockHTML( { guid: GUID_INITIAL_VIDEO } ),
		} );
		const { getByLabelText, getByTestId } = screen;

		// Block is visible
		const block = await getBlock( screen, 'VideoPress' );
		expect( block ).toBeVisible();
		fireEvent.press( block );

		expect( getEditorHtml() ).toMatchSnapshot( 'before replacing video' );

		// Setup pickers
		const { expectMediaPickerCall, mediaPickerCallback } =
			setupMediaPicker();
		const { selectOption } = setupPicker( screen, MEDIA_OPTIONS );

		// Mock API responses for new block
		setupApiFetch(
			generateFetchMocks( {
				guid: GUID_REPLACE_VIDEO,
				metadata: {
					title: 'Video from media library',
				},
			} )
		);

		// Replace video
		fireEvent.press( getByLabelText( 'Edit video' ) );

		// Add video from WordPress media library
		selectOption( 'WordPress Media Library' );
		expectMediaPickerCall( 'SITE_MEDIA_LIBRARY', [ 'video' ], false );
		await mediaPickerCallback(
			generateLibraryMedia( { guid: GUID_REPLACE_VIDEO } )
		);

		expect( getByTestId( 'videopress-player' ) ).toBeVisible();
		expect( getEditorHtml() ).toMatchSnapshot( 'after replacing video' );
	} );

	it( 'replaces with new URL', async () => {
		const screen = await initializeEditor( {
			initialHtml: generateBlockHTML( { guid: GUID_INITIAL_VIDEO } ),
		} );
		const {
			getByLabelText,
			getByTestId,
			getByPlaceholderText,
			findByTestId,
		} = screen;

		// Block is visible
		const block = await getBlock( screen, 'VideoPress' );
		expect( block ).toBeVisible();
		fireEvent.press( block );

		expect( getEditorHtml() ).toMatchSnapshot( 'before replacing video' );

		// Setup pickers
		const { selectOption } = setupPicker( screen, MEDIA_OPTIONS );

		// Mock API responses for new block
		setupApiFetch(
			generateFetchMocks( {
				guid: GUID_REPLACE_VIDEO,
				metadata: {
					title: 'Video from URL',
				},
			} )
		);

		// Replace video
		fireEvent.press( getByLabelText( 'Edit video' ) );

		// Insert video from URL
		selectOption( 'Insert from URL' );
		fireEvent.changeText(
			getByPlaceholderText( 'Type a URL' ),
			`https://videopress.com/v/${ GUID_REPLACE_VIDEO }`
		);
		dismissModal( getByTestId( 'bottom-sheet' ) );
		sendWebViewMessage( await findByTestId( 'videopress-player' ), {
			type: 'message',
			event: 'videopress_ready',
		} );

		expect( getByTestId( 'videopress-player' ) ).toBeVisible();
		expect( getEditorHtml() ).toMatchSnapshot( 'after replacing video' );
	} );

	it( 'restores previous video when canceling a video replacement', async () => {
		const screen = await initializeEditor( {
			initialHtml: generateBlockHTML( { guid: GUID_INITIAL_VIDEO } ),
		} );
		const { getByLabelText, getByTestId } = screen;

		// Block is visible
		const block = await getBlock( screen, 'VideoPress' );
		expect( block ).toBeVisible();
		fireEvent.press( block );

		expect( getEditorHtml() ).toMatchSnapshot( 'before replacing video' );

		// Setup pickers and media upload
		const { notifyUploadingState, notifyResetState } = setupMediaUpload();
		const { expectMediaPickerCall, mediaPickerCallback } =
			setupMediaPicker();
		const { selectOption } = setupPicker( screen, MEDIA_OPTIONS );

		// Mock API responses for new block
		setupApiFetch(
			generateFetchMocks( {
				guid: GUID_REPLACE_VIDEO,
				metadata: {
					title: 'Local video from device',
				},
			} )
		);

		// Replace video
		fireEvent.press( getByLabelText( 'Edit video' ) );

		// Upload video from device
		selectOption( 'Choose from device' );
		expectMediaPickerCall( 'DEVICE_MEDIA_LIBRARY', [ 'video' ], false );
		const localMedia = generateLocalMedia();
		await mediaPickerCallback( localMedia );
		await notifyUploadingState( localMedia );

		// During upload progress we keep displaying the loading state
		const uploadingVideoView = getByTestId( 'videopress-uploading-video' );
		expect( uploadingVideoView ).toBeVisible();

		// Restores mocks for API requests of initial video
		setupApiFetch(
			generateFetchMocks( {
				guid: GUID_INITIAL_VIDEO,
				metadata: {
					title: 'Video to be replaced',
				},
			} )
		);
		// Cancel upload
		fireEvent.press( uploadingVideoView );
		expect( requestImageUploadCancelDialog ).toHaveBeenCalledWith(
			localMedia.localId
		);
		await notifyResetState( localMedia );

		expect( getByTestId( 'videopress-player' ) ).toBeVisible();
		expect( getEditorHtml() ).toMatchSnapshot(
			'after canceling replacement'
		);
	} );
} );
