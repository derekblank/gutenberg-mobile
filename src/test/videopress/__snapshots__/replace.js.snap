// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPress block - Replace replaces with a local video: after replacing video 1`] = `"<!-- wp:videopress/video {"title":"Local video from device","description":"video-description","useAverageColor":false,"id":34,"guid":"NEWxGUID","privacySetting":2,"allowDownload":false,"rating":"G","isPrivate":false,"duration":1200} /-->"`;

exports[`VideoPress block - Replace replaces with a local video: before replacing video 1`] = `"<!-- wp:videopress/video {"title":"Video to be replaced","description":"video-description","useAverageColor":false,"id":34,"guid":"OLDxGUID","privacySetting":2,"allowDownload":false,"rating":"G","isPrivate":false,"duration":1200} /-->"`;

exports[`VideoPress block - Replace replaces with new URL: after replacing video 1`] = `"<!-- wp:videopress/video {"title":"Video from URL","description":"video-description","useAverageColor":false,"id":34,"guid":"NEWxGUID","src":"https://videopress.com/v/NEWxGUID","privacySetting":2,"allowDownload":false,"rating":"G","isPrivate":false,"duration":1200} /-->"`;

exports[`VideoPress block - Replace replaces with new URL: before replacing video 1`] = `"<!-- wp:videopress/video {"title":"Video to be replaced","description":"video-description","useAverageColor":false,"id":34,"guid":"OLDxGUID","privacySetting":2,"allowDownload":false,"rating":"G","isPrivate":false,"duration":1200} /-->"`;

exports[`VideoPress block - Replace replaces with video from media library: after replacing video 1`] = `"<!-- wp:videopress/video {"title":"Video from media library","description":"video-description","useAverageColor":false,"id":34,"guid":"NEWxGUID","privacySetting":2,"allowDownload":false,"rating":"G","isPrivate":false,"duration":1200} /-->"`;

exports[`VideoPress block - Replace replaces with video from media library: before replacing video 1`] = `"<!-- wp:videopress/video {"title":"Video to be replaced","description":"video-description","useAverageColor":false,"id":34,"guid":"OLDxGUID","privacySetting":2,"allowDownload":false,"rating":"G","isPrivate":false,"duration":1200} /-->"`;

exports[`VideoPress block - Replace restores previous video when canceling a video replacement: after canceling replacement 1`] = `"<!-- wp:videopress/video {"title":"Video to be replaced","description":"video-description","useAverageColor":false,"id":34,"guid":"OLDxGUID","src":"file:///local-video-1.mp4","privacySetting":2,"allowDownload":false,"rating":"G","isPrivate":false,"duration":1200} /-->"`;

exports[`VideoPress block - Replace restores previous video when canceling a video replacement: before replacing video 1`] = `"<!-- wp:videopress/video {"title":"Video to be replaced","description":"video-description","useAverageColor":false,"id":34,"guid":"OLDxGUID","privacySetting":2,"allowDownload":false,"rating":"G","isPrivate":false,"duration":1200} /-->"`;

exports[`VideoPress block - Replace takes a new video: after replacing video 1`] = `"<!-- wp:videopress/video {"title":"Recorded within the editor","description":"video-description","useAverageColor":false,"id":34,"guid":"NEWxGUID","privacySetting":2,"allowDownload":false,"rating":"G","isPrivate":false,"duration":1200} /-->"`;

exports[`VideoPress block - Replace takes a new video: before replacing video 1`] = `"<!-- wp:videopress/video {"title":"Video to be replaced","description":"video-description","useAverageColor":false,"id":34,"guid":"OLDxGUID","privacySetting":2,"allowDownload":false,"rating":"G","isPrivate":false,"duration":1200} /-->"`;
